internal/data/common.go

```diff
@@ -126,7 +128,161 @@ func (a AttributeValue) SetBool(v bool) {
+func NewAttributeKeyValueString(k string, v string) AttributeKeyValue {
+	akv := AttributeKeyValue{&otlpcommon.AttributeKeyValue{Key: k}}
+	akv.Value().SetString(v)
+	return akv
+}
+
+// NewAttributeKeyValueInt creates a new AttributeKeyValue with the given key and int64 value.
+func NewAttributeKeyValueInt(k string, v int64) AttributeKeyValue {
+	akv := AttributeKeyValue{&otlpcommon.AttributeKeyValue{Key: k}}
+	akv.Value().SetInt(v)
+	return akv
+}
+
+// NewAttributeKeyValueDouble creates a new AttributeKeyValue with the given key and float64 value.
+func NewAttributeKeyValueDouble(k string, v float64) AttributeKeyValue {
+	akv := AttributeKeyValue{&otlpcommon.AttributeKeyValue{Key: k}}
+	akv.Value().SetDouble(v)
+	return akv
+}
+
+// NewAttributeKeyValueBool creates a new AttributeKeyValue with the given key and bool value.
+func NewAttributeKeyValueBool(k string, v bool) AttributeKeyValue {
+	akv := AttributeKeyValue{&otlpcommon.AttributeKeyValue{Key: k}}
+	akv.Value().SetBool(v)
+	return akv
+}
+
+// Key returns the key associated with this AttributeKeyValue.
+func (akv AttributeKeyValue) Key() string {
+	return akv.orig.Key
+}
+
+// Value returns the value associated with this AttributeKeyValue.
+func (akv AttributeKeyValue) Value() AttributeValue {
+	return AttributeValue{akv.orig}
+}
+
+// SetValue replaces the value associated with this AttributeKeyValue.
+func (akv AttributeKeyValue) SetValue(av AttributeValue) {
+	akv.orig.Type = av.orig.Type
+	akv.orig.StringValue = av.orig.StringValue
+	akv.orig.IntValue = av.orig.IntValue
+	akv.orig.DoubleValue = av.orig.DoubleValue
+	akv.orig.BoolValue = av.orig.BoolValue
+}
+
+// AttributeMap stores a map of attribute keys to values.
+type AttributeMap struct {
+	orig *[]*otlpcommon.AttributeKeyValue
+}
+
+// NewAttributeMap creates a new AttributeMap from the given map[string]AttributeValue.
+func NewAttributeMap(attrMap map[string]AttributeValue) AttributeMap {
+	origs := make([]otlpcommon.AttributeKeyValue, len(attrMap))
+	wrappers := make([]*otlpcommon.AttributeKeyValue, len(attrMap))
+
+	ix := 0
+	for k, v := range attrMap {
+		wrappers[ix] = &origs[ix]
+		wrappers[ix].Key = k
+		AttributeKeyValue{wrappers[ix]}.SetValue(v)
+		ix++
+	}
+
+	return AttributeMap{&wrappers}
+}
+
+// Get returns the AttributeKeyValue associated with the key and true,
+// otherwise an invalid instance of the AttributeKeyValue and false.
+func (am AttributeMap) Get(key string) (AttributeKeyValue, bool) {
+	for _, a := range *am.orig {
+		if a.Key == key {
+			return AttributeKeyValue{a}, true
+		}
+	}
+	return AttributeKeyValue{nil}, false
+}
+
+// Delete deletes the entry associated with the key and returns true if the key
+// was present in the map, otherwise returns false.
+func (am AttributeMap) Delete(key string) bool {
+	for i, a := range *am.orig {
+		if a.Key == key {
+			(*am.orig)[i] = (*am.orig)[len(*am.orig)-1]
+			*am.orig = (*am.orig)[:len(*am.orig)-1]
```

Comment:
nit: this shouldn't make any practical difference due to the way that they are allocated but in the unlike case that the pointer put in index `i` is later released adding `(*am.orig)[len(*am.orig)-1] = nil` ensures no lingering references from the part of the slice that is not used anymore but still on memory.